File: packages/llm/src/retry.ts
export async function withRetries<T>(
  fn: () => Promise<T>,
  opts: { attempts: number; baseMs: number; maxMs: number },
): Promise<T> {
  let last: unknown;
  for (let i = 0; i < opts.attempts; i += 1) {
    try {
      return await fn();
    } catch (e) {
      last = e;
      if (i === opts.attempts - 1) break;
      const delay = Math.min(
        opts.maxMs,
        opts.baseMs * 2 ** i + Math.floor(Math.random() * 100),
      );
      await new Promise((r) => setTimeout(r, delay));
    }
  }
  throw last;
}

File: rag-service/rag_service/llm/retry.py
from __future__ import annotations

import random
import time
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")


def with_retries(
    fn: Callable[[], T],
    *,
    attempts: int,
    base_ms: int,
    max_ms: int,
) -> T:
    last: BaseException | None = None
    for i in range(attempts):
        try:
            return fn()
        except BaseException as e:
            last = e
            if i == attempts - 1:
                break
            delay_ms = min(
                max_ms,
                base_ms * (2**i) + random.randint(0, 99),
            )
            time.sleep(delay_ms / 1000.0)
    assert last is not None
    raise last
